packages/core/src/lib/connection_manager.ts

```diff
@@ -44,12 +52,57 @@ export class ConnectionManager {
     return instance;
   }
 
+  public async getPeersByDiscovery(): Promise<PeersByDiscovery> {
+    const peersDiscovered = await this.libp2p.peerStore.all();
+    const peersConnected = this.libp2p
+      .getConnections()
+      .map((conn) => conn.remotePeer);
+
+    const peersDiscoveredByBootstrap: Peer[] = [];
+    const peersDiscoveredByPeerExchange: Peer[] = [];
+    const peersConnectedByBootstrap: Peer[] = [];
+    const peersConnectedByPeerExchange: Peer[] = [];
+
+    for (const peer of peersDiscovered) {
+      const tags = await this.getTagNamesForPeer(peer.id);
```

Comment:
you don't need to use `getTagNamesForPeer` but you can directly refer to `peer.tags.keys()` to avoid async operation

Reply:
I am wondering what is the difference if the same `tags` are present on `peer` in the loop as well as those you'd get from `this.getTagNamesForPeer`?